### ISSUE TITLE
Securing togglz dashboard and read-only mode?

### ISSUE THREAD

**Issue by alecl** [togglz/togglz#205]
I'm looking to secure the dashboard. Was considering putting some Tomcat user auth in front of it but ideally would like to separate who can view values vs change or outright disable change from the interface. Is that something supported in some manner already or would need to be created?

**Comment by chkal** [comment]
Hey @alecl,

currently it is only possible to either allow or deny access to the Togglz Console via `FeatureUser#isFeatureAdmin()`. I'm not sure if it makes sense to separate this into separate permissions to be honest. But of cause you could use Tomcat user authentication for this.

Christian

**Comment by alecl** [comment]
Most companies want stricter controls and auditing on who can change a setting vs who can view it so the separate permissions would be welcome.

**Comment by chkal** [comment]
I agree that more fine grained control would be a good thing. However, in those cases you typically also have many other requirements which cannot be provided by the default console. Therefore implementing a custom web UI for toggling features is a good alternative.

However, PRs for implementing something like this are more than welcome.

**Comment by charithdesilva** [comment]
hi @chkal I'm looking to implement this as part of my requirement to have fine grained control. If any of you have done any analysis on this already, that'd be helpful for me to get started. My key requirement is to limit control on togglz based on user's role.

**Comment by chkal** [comment]
Hey @charithdesilva. Nice to hear that you would like to work on this issue.

The big question is how to tell Togglz that a user has readonly vs read-write permission. Currently there is just `FeatureUser#isFeatureAdmin()`. I've no idea how to implement this in a backward-compatible way without breaking interfaces.

**Comment by emahaletmc** [comment]
Hey @chkal, Kudos to you for coming up with Togglz!  

I was trying to look at a possibility of incorporating this into my application. One of my concern is giving access to anyone to come and edit the feature toggle which got me thinking.
I was thinking of adding something like "isReadOnly" which would be defaulted to false to ensure backward compatibility and can be controlled with FeatureUser#isReadOnly(). What do you think? --  Hide the action tab based on that flag, which would be passed through to the feature user. Just a hypothesis. Trying to find a way to solve this.

**Comment by chkal** [comment]
Hey @emahaletmc. Thanks for sharing you thoughts. I'm not sure if something like 'isReadOnly()` would be flexible enough. Also adding a new method to the interface will definitely break implementations, which is a huge problem. I think for Togglz 3 we need something even more powerful. Something like `public AccessLevel getPermission(Feature feature)`.

**Comment by emahaletmc** [comment]
@chkal, Do you have something in mind for Togglz 3? It'd be really great to have this be a part of it. 

If so, when would Togglz 3 be targetted for? 

Eshan

**Comment by chkal** [comment]
Hey @emahaletmc ,

to be honest, there is no clear vision or a target for Togglz 3 at the moment. I don't find much time anymore to work on Togglz. It has always been a spare time project. And being a father of two there is not much spare time anymore. I already thought about stepping back as a project lead, but I fear that the project will die in this case. 

Therefore, I started a [OpenCollective](https://opencollective.com/togglz/) group for Togglz to allow users/companies to support the project. However, there wasn't much interest in supporting Togglz till now. Which is very disappointing with more than 60k artifact downloads on Maven Central per month.

Maybe I'll get more support at some point. Not sure. I'm currently in a very demotivated mood about spending more time on the project. Especially because of the lack of interest of the community in supporting it.
